### PR TITLE
cgroupv2: fix fs2 driver initialization

### PR DESCRIPTION
**Pull request by kolyshkin**
The main goal of this patchset is to fix fs2 driver initialization. With these patches, the tests that I added in https://github.com/opencontainers/runc/pull/2295 are now passing with fs2 driver (i.e. when `RUNC_USE_SYSTEMD` is not set), except for one case (https://github.com/opencontainers/runc/issues/2298).

I also did some major cleanups, mostly for systemd driver. A high-level overview:
 - remove legacy paths map from v2 driver
 - untangle systemd v2 initialization;
 - reorganized systemd code into v1, v2, common, and unsupported.

Please see individual commits for details.

@AkihiroSuda PTAL (this was discussed before in https://github.com/opencontainers/runc/pull/2281#issuecomment-609099674)

### CODE REVIEW

File: libcontainer/cgroups/fs2/fs2.go [hunk 1]
```diff
@@ -8,75 +8,67 @@ import (
...
-	securejoin "github.com/cyphar/filepath-securejoin"
 	"github.com/opencontainers/runc/libcontainer/cgroups"
 	"github.com/opencontainers/runc/libcontainer/configs"
 	"github.com/pkg/errors"
 )
 
+type manager struct {
+	config *configs.Cgroup
+	// dirPath is like "/sys/fs/cgroup/user.slice/user-1001.slice/session-1.scope"
+	dirPath string
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/fs2.go:19]: This example is confusing and looks as if it was related to rootless systemd

**Comment by kolyshkin** [comment]
> `write /sys/fs/cgroup/cgroup.subtree_control: device or resource busy`

* if this is in test with `--systemd-cgroup`, then I don't understand why is this happening as I haven't changed anything;
* if this is in test without `--systemd-cgroup` but systemd is present, then `--systemd-cgroup` should be used;
* if this is in test without `--systemd-cgroup` and systemd is not present, then I don't have an idea about what is going on :(

**Comment by AkihiroSuda** [comment]
the errors are happening on Travis
https://travis-ci.org/github/opencontainers/runc/jobs/672264460

**Comment by kolyshkin** [comment]
The problem is we're enabling all the controllers no matter what limits are set. Two ways to fix this are:

1. enable all controllers on a "best effort" basis (i.e. ignore the errors during init time, return errors during set time)
2. only enable needed controllers.

I like 2 much better. Let's see if this can be done (and it will fix the above tests since looks like they do not set any limits)

**Comment by kolyshkin** [comment]
Added a commit to enable controllers only when required and on a best-effort basis.

cgroup-v2 is passing now. The tests I add/enable in #2295 are passing, too, on my test box.

**Comment by kolyshkin** [comment]
@crosbymichael @mrunalp PTAL

**Comment by kolyshkin** [comment]
@cyphar PTAL (at least for those two patches that add/remove securejoin usage)

**Comment by kolyshkin** [comment]
Rebased to resolve conflict after #2288 is merged

**Comment by kolyshkin** [comment]
Let me rework the last patch (enable controllers on demand) -- while it works fine, I don't like that `neededControllers()` has the intimate knowledge of all subsystems in one place.

**Comment by kolyshkin** [comment]
@AkihiroSuda please let me know if you want me to split into two PRs, as this one might be too big to review. I tried hard to split it into small logical commits but there are too many.

**Comment by kolyshkin** [comment]
fixed a bug in last commit

File: libcontainer/cgroups/fs2/cpu.go [hunk 2]
```diff
@@ -13,7 +13,15 @@ import (
 	"github.com/opencontainers/runc/libcontainer/configs"
 )
 
+func isCpuSet(cgroup *configs.Cgroup) bool {
+	return cgroup.Resources.CpuWeight != 0 || cgroup.Resources.CpuMax != ""
+}
+
 func setCpu(dirPath string, cgroup *configs.Cgroup) error {
+	if !isCpuSet(cgroup) {
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/cpu.go:21]: Not sure whether this validation is useful.

File: libcontainer/cgroups/fs2/cpuset.go [hunk 3]
```diff
@@ -7,7 +7,15 @@ import (
 	"github.com/opencontainers/runc/libcontainer/configs"
 )
 
+func isCpusetSet(cgroup *configs.Cgroup) bool {
+	return cgroup.Resources.CpusetCpus != "" || cgroup.Resources.CpusetMems != ""
+}
+
 func setCpuset(dirPath string, cgroup *configs.Cgroup) error {
+	if !isCpusetSet(cgroup) {
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/cpuset.go:15]: same as above

File: libcontainer/cgroups/fs2/create.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/opencontainers/runc/libcontainer/configs"
+)
+
+func neededControllers(cgroup *configs.Cgroup) []string {
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/create.go:13]: Please add godoc that the list doesn't contain "pseudo" controllers, i.e. devices and freezer

File: libcontainer/cgroups/fs2/create.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	var list []string
+
+	if cgroup == nil {
+		return list
+	}
+
+	// list of all available controllers
+	const file = UnifiedMountpoint + "/cgroup.controllers"
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/fs2/create.go:21]: TODO (#2281):  this should be `/sys/fs/cgroup/user.slice/user-1001.slice/cgroup.controllers` in rootless systemd

File: libcontainer/cgroups/fs2/cpu.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/cpu.go:21]: The reason I added it is peculiar -- if someone will be modifying `setCpu()`, adding more fields to set, they will be reminded to also modify `isCpuSet()`. Having this check here _might_ also help to uncover bugs.

File: libcontainer/cgroups/systemd/common.go [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return false
+	}
+	return fi.IsDir()
+}
+
+// systemd represents slice hierarchy using `-`, so we need to follow suit when
+// generating the path of slice. Essentially, test-a-b.slice becomes
+// /test.slice/test-a.slice/test-a-b.slice.
+func ExpandSlice(slice string) (string, error) {
```
Comment (AkihiroSuda) [inline libcontainer/cgroups/systemd/common.go:38]: Not new in this PR, but can we have UT for this?

File: libcontainer/cgroups/fs2/cpu.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/cpu.go:21]: E.g. someone adds `cpuFoobar` setting to setCpu(), but forgets to modify `isCpuSet()`. They add a unit test to check if `cpuFoobar` is applied, and the test fails! Hope it makes sense.

File: libcontainer/factory_linux.go [hunk 7]
```diff
@@ -50,28 +50,51 @@ func InitArgs(args ...string) func(*LinuxFactory) error {
...
+			panic(errors.Errorf("invalid dir path %q", path))
+		}
+	}
+
+	return path
+}
+
+func systemdCgroupV2(l *LinuxFactory, rootless bool) error {
+	l.NewCgroupsManager = func(config *configs.Cgroup, paths map[string]string) cgroups.Manager {
+		return systemd.NewUnifiedManager(config, getUnifiedPath(paths), rootless)
```
Comment (AkihiroSuda) [inline libcontainer/factory_linux.go:74]: Can be another PR, but eventually we should remove `getUnifiedPath()` madness

File: libcontainer/cgroups/fs2/cpu.go (hunk 2 above)
Comment (AkihiroSuda) [reply libcontainer/cgroups/fs2/cpu.go:21]: Can be just code comment lines?

File: libcontainer/cgroups/fs2/create.go (hunk 5 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/create.go:21]: I'm not sure. We're getting the list of controllers available on the host, it does not matter rootless or not. Even if some controllers are not available for the rootless cgroup, we won't error out, so having them in this list is OK.

File: libcontainer/factory_linux.go (hunk 7 above)
Comment (kolyshkin) [reply libcontainer/factory_linux.go:74]: Yes, totally agree, I plan to do that, but this PR is way too big already, let's do it step by step and one day we'll get there :)

File: libcontainer/cgroups/fs2/create.go (hunk 4 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/create.go:13]: done

File: libcontainer/cgroups/fs2/cpu.go (hunk 2 above)
Comment (kolyshkin) [reply libcontainer/cgroups/fs2/cpu.go:21]: > Can be just code comment lines?

What do you mean? Add a comment like "if you're modifying this, don't forget to modify isSetCpu"? In my experience, this never works. The added runtime check might actually help to uncover such a bug.

Or do you mean adding a comment stating the reason why this check is added?

Comment (kolyshkin) [reply libcontainer/cgroups/fs2/cpu.go:21]: > Or do you mean adding a comment stating the reason why this check is added?

I have amended the commit telling why is*Set is added to set*:


4. Amend all setFoo() functions to call isFooSet(). While this might
   seem unnecessary, it might actually help to uncover a bug.
   Imagine someone:
    - adds a cgroup.Resources.CpuFoo setting;
    - modifies setCpu() to apply the new setting;
    - but forgets to amend isCpuSet() accordingly <-- BUG

   In this case, a test case modifying CpuFoo will help
   to uncover the BUG. This is the reason why it's added.

Comment (AkihiroSuda) [reply libcontainer/cgroups/fs2/cpu.go:21]: I mean the former, but not a blocker

**Comment by AkihiroSuda** [comment]
Looks good, but I'd like to merge the integration tests (https://github.com/opencontainers/runc/pull/2295) before this.

File: libcontainer/cgroups/systemd/common.go (hunk 6 above)
Comment (kolyshkin) [reply libcontainer/cgroups/systemd/common.go:38]: Will do (maybe in a separate PR?)

Comment (AkihiroSuda) [reply libcontainer/cgroups/systemd/common.go:38]: Yes, can be a separate PR

**Comment by kolyshkin** [comment]
Rebasing since #2295 is merged. I still need #2305 otherwise the tests will definitely fail.

**Comment by kolyshkin** [comment]
- rebased on top of #2295
 - (temporarily) cherry-picked #2305
 - enabled the localintegration test

**Comment by kolyshkin** [comment]
checkpoint/restore tests fail; let me try without

**Comment by kolyshkin** [comment]
OK this breaks checkpoint for some reason. Looking

**Comment by kolyshkin** [comment]
Disabled checkpoint tests (runc currently does not support cgroup v2 c/r). All others should pass

**Comment by kolyshkin** [comment]
Something ugly with CI -- logs for failed jobs are empty. Restarting

**Comment by kolyshkin** [comment]
This one is ready for merge.

@crosbymichael @mrunalp @filbranden PTAL

**Comment by AkihiroSuda** [comment]
LGTM

**Comment by AkihiroSuda** [comment]
LGTM (still)

**Comment by mrunalp** [comment]
LGTM

**Comment by AkihiroSuda** [comment]
This commit caused regression https://github.com/opencontainers/runc/issues/2339
